crates/matrix-sdk/src/room_preview.rs

```diff
@@ -254,101 +250,181 @@ impl RoomPreview {
     }
 }
 
-pub enum RoomPreviewActions {
-    Invited { join: JoinRoomPreviewAction, leave: LeaveRoomPreviewAction },
-    Knocked { leave: LeaveRoomPreviewAction },
+/// A set of actions that can be performed in a [`RoomPreview`].
+#[derive(Debug)]
+pub struct RoomPreviewActions {
```

Comment:
Overall, I don't think this data structure and the other actions are required at the SDK level; maybe they're required at the FFI layer because we need some kind of "lazy provider"-like object for the Record requirements, but I don't think they make sense for the `RoomPreview` of the SDK, considering there are many other methods that already exist providing the same functionality, so removing it would avoid blowing up the number of abstractions, code duplication, etc.

Reply:
I added it to the SDK so they could be tested and didn't have to duplicate code, i.e. the `Room::join` contents mentioned below, which could be just put into one of these actions and then reused inside `Room::join` instead:

```rust
impl Room {
    async fn join(&self) -> Result<_, _> {
        self.actions.join()
   }
```

But if that's not something we want, we can just remove the actions, leaving a needed `RoomPreview::leave` method that's needed (`Client` has code for joining and knocking, but it doesn't have one for leaving a room given its id, so a `RoomPreview` with invite/knock state can never be left like that).